server/ladder_service.py

```diff
@@ -27,39 +35,80 @@ def inform_player(self, player):
-        game.set_player_option(player2.id, 'Army', 3)
+        game.set_player_option(host.id, 'StartSpot', 1)
+        game.set_player_option(guest.id, 'StartSpot', 2)
+        game.set_player_option(host.id, 'Faction', host.faction)
+        game.set_player_option(guest.id, 'Faction', guest.faction)
+        game.set_player_option(host.id, 'Color', 1)
+        game.set_player_option(guest.id, 'Color', 2)
+        game.set_player_option(host.id, 'Army', 2)
+        game.set_player_option(guest.id, 'Army', 3)
 
         # Remembering that "Team 1" corresponds to "-": the non-team.
-        game.set_player_option(player1.id, 'Team', 1)
-        game.set_player_option(player2.id, 'Team', 1)
+        game.set_player_option(host.id, 'Team', 1)
+        game.set_player_option(guest.id, 'Team', 1)
 
         mapname = map_path[5:-4]  # FIXME: Database filenames contain the maps/ prefix and .zip suffix.
                                   # Really in the future, just send a better description
-        player1.lobby_connection.launch_game(game, player1.game_port, is_host=True, use_map=mapname)
+        host.lobby_connection.launch_game(game, host.game_port, is_host=True, use_map=mapname)
         await asyncio.sleep(4)
-        player2.lobby_connection.launch_game(game, player2.game_port, is_host=False, use_map=mapname)
+        guest.lobby_connection.launch_game(game, guest.game_port, is_host=False, use_map=mapname)
+
+    async def choose_map(self, players: [Player]) -> Tuple[int, str, str]:
+        maps = self.game_service.ladder_maps
+
+        if not maps:
+            self._logger.error("Trying to choose a map from an empty map pool!")
+            raise RuntimeError("Ladder maps not set!")
+
+        recently_played_map_ids = {
+            map_id for player in players for map_id, _ in
+            await self.get_ladder_history(player)
+        }
+        randomized_maps = random.sample(maps, len(maps))
+
+        try:
+            return next(
+                filter(
+                    lambda m: m[0] not in recently_played_map_ids,
+                    randomized_maps
+                )
+            )
+        except StopIteration:
+            # If all maps were played recently, pick a random one
+            return randomized_maps[0]
+
+    async def get_ladder_history(self, player: Player, limit=3) -> List[Tuple[int, datetime]]:
+        async with db.engine.acquire() as conn:
+            query = select([
+                game_stats.c.mapId,
+                game_player_stats.c.scoreTime
+            ]).select_from(
+                game_player_stats.join(game_stats).join(game_featuredMods)
+            ).where(
+                and_(
+                    game_player_stats.c.playerId == player.id,
+                    game_player_stats.c.scoreTime >= func.now() - text("interval 1 day"),
+                    game_featuredMods.c.gamemod == "ladder1v1"
+                )
+            ).limit(limit)
```

Comment:
If you'll change models, i'll help you write query in other way.

ala:
```
conn.query(GameStats.mapId).select_from(
    GamePlayerStats
).join(
    GameStats, 
    GamePlayerStats.gameId == GameStats.id
).join(
    GameFeaturedMods,
    GameFeaturedMods.id == GameStats.gameMod
).where(
    ...
)
```

SQLAlchemy accepts functions `from sqlalchemy import func`

Better example is here: https://stackoverflow.com/questions/30495935/flask-sqlalchemy-query-datetime-intervals